### ISSUE TITLE
listen along

### ISSUE THREAD

**Issue by Hooooooman** [th-ch/youtube-music#992]
How do I listen along on discord without opening browser and redirect the link to the app seamlessly

**Comment by Zo-Bro-23** [comment]
Currently, you can't, but it is possible to add a feature that does this using App URLs or `xdg-open` (`start` in Windows) so that you see the "Open with X app?" prompt in the browser (or perhaps even directly; not sure if that is possible though).

@Araxeus interested?

**Comment by Araxeus** [comment]
This is actually a good idea!
After #946 we might be able to directly launch youtube music with the new protocol.

Just have to add a few new commands like `load the url` which is currently not possible on song-controls

**Comment by Zo-Bro-23** [comment]
> That's an interesting idea...
> 
> After #946 we might be able to directly launch youtube music with the new protocol.
> 
> Just have to add a few new commands like `load the url` which is currently not possible on song-controls
> 
> The problem is:
> 
> If a user click the `listen along` button and doesn't have youtube-music installed - **it won't do anything**
> 
> Considering the fact above, I'm not sure how to do it

A browser redirect would probably be best in that case. If YTM is installed, it should prompt to open the song in YTM; otherwise, it can redirect to https://music.youtube.com.

Actually, doing that for all https://music.youtube.com URLs would be cool! We can have a browser extension that redirects ***all*** YTM URLs to the app!

**Comment by Hooooooman** [comment]
> That's an interesting idea...
> 
> After #946 we might be able to directly launch youtube music with the new protocol.
> 
> Just have to add a few new commands like `load the url` which is currently not possible on song-controls
> 
> The problem is:
> 
> If a user click the `listen along` button and doesn't have youtube-music installed - **it won't do anything**
> 
> Considering the fact above, I'm not sure how to do it

Yea I think it would be a good addition to the app, many of my friends tends to use youtube music because of this repository but it lacks the option to listen along like spotify, I personally like youtube music(especially this desktop version) because of comparatively better recommendations and the ability to watch music video without opening the browser. I really hope this idea turns into a feature soon :)

**Comment by xamionex** [comment]
Another idea i just came up for discord directly opening ytm is to make a second button, I know you can make the rich presences have 2 buttons so you could make a 2nd one saying "`Open in YTMusic`/`YTMusic App`". This could save users with the app some time with having the app open the link directly and no browser launching

**Comment by Rairof** [comment]
Just putting this out here as a referenced feature example picture for others to see-
![1701613784204](https://github.com/th-ch/youtube-music/assets/107706560/a945cb5d-7bf1-490f-887e-07ceac48984e)
I'm on board with this idea, hope it gets created soon enough :)

**Comment by JacKooDesu** [comment]
I'm trying to build msix for ytm lately
Struggling with bundle signing though, I think build the project in msix instead of nsis for win platform can achieve this feature. 
By configuring `appxmanifest.xml` which can set default application to open url.

p.s. I don't familiar with other platforms neither linux distros or macOs. Should only possible achieve by using protocol register I thought?

**Comment by xamionex** [comment]
URIs on linux are added like this:
1. Add MimeType to your .desktop file for the application
   - `MimeType=x-scheme-handler/youtube-music;`
2. Use as the default app for URI
   - `xdg-mime default youtube-music.desktop x-scheme-handler/youtube-music`
   - `update-desktop-database`

**Comment by ArjixWasTaken** [comment]
We already have a custom protocol...

**Comment by xamionex** [comment]
> We already have a custom protocol...

My bad.

**Comment by JacKooDesu** [comment]
> URIs on linux are added like this:
> 
>     1. Add MimeType to your .desktop file for the application
>        
>        * `MimeType=x-scheme-handler/youtube-music;`
> 
>     2. Use as the default app for URI
>        
>        * `xdg-mime default youtube-music.desktop x-scheme-handler/youtube-music`
>        * `update-desktop-database`

yep, the `youtubemusic://` protocol already added for all platform i guess.

bundling in `.msix` not only provide feature to open application from url, additionally, updating in a better way and run application in sandbox are contained.
also, `.msix` can be uploaded to ms store, which provide user who is not familiar using github an another way to download the application.

though, there still some limitations like cannot install properly if user is not using win10 or later.

btw, is there anyone who has experience in bundling `.msix`? i can bundle whole execution binary but still encountered errors when signing lol.

**Comment by JacKooDesu** [comment]
Windows can config custom protocol by registry files. You can simply check it in regedit (usually in `HKEY_CLASSES_ROOT/[AppName]/Url Protocol`)
Msix side is another way kind like deep link on Android and cannot config by user.

**Comment by ArjixWasTaken** [comment]
> I'm sorry if I missed something, but how to do that on Windows?
> - Interesting note - I tried to use `youtubemusic://watch?v=VIDEO_ID` - app opens, but ignores the song/video in the link.

Because that's incorrect

**Comment by JacKooDesu** [comment]
> > I'm sorry if I missed something, but how to do that on Windows?
> > - Interesting note - I tried to use `youtubemusic://watch?v=VIDEO_ID` - app opens, but ignores the song/video in the link.
> 
> Because that's incorrect 

Reference `SongInfo.url` which is provided to create discord button

**Comment by bennett-sh** [comment]
> I'm sorry if I missed something, but how to do that on Windows?
> 
>     * Interesting note - I tried to use `youtubemusic://watch?v=VIDEO_ID` - app opens, but ignores the song/video in the link.

Is it just not intended for deep links to receive arguments or is there a valid syntax? I've looked through the code but it doesn't seem like so (and I'd like to open the search page from an external program).

**Comment by BillCipher345** [comment]
> > I'm sorry if I missed something, but how to do that on Windows?
> > 
> > * Interesting note - I tried to use `youtubemusic://watch?v=VIDEO_ID` - app opens, but ignores the song/video in the link.
> 
> Because that's incorrect

could you please explain how is this incorrect and how to use the protocol?

**Comment by bennett-sh** [comment]
> > > I'm sorry if I missed something, but how to do that on Windows?
> > > 
> > > * Interesting note - I tried to use `youtubemusic://watch?v=VIDEO_ID` - app opens, but ignores the song/video in the link.
> > 
> > 
> > Because that's incorrect
> 
> could you please explain how is this incorrect and how to use the protocol?

Generally the protocol works like so: `youtubemusic://[method name] [...args]`. Everything after the protocol should be url encoded (i.e. `addSongToQueue ID` -> `addSongToQueue%20ID`).

Examples:
`youtubemusic://play`
`youtubemusic://muteUnmute`
`youtubemusic://addSongToQueue%20SOME_VIDEO_ID%20INSERT_AFTER_CURRENT_VIDEO`

Here you can find all the methods available (some of them don't work for me though):
https://github.com/pear-devs/pear-desktop/blob/b37618779eec5cfd42d2b7b12c1761c95874aa11/src/providers/song-controls.ts#L39

If you want to play a video/song, the (to my knowledge) best approach would be a combination of these:
1. Add Song to Queue: `youtubemusic://addSongToQueue%20SONG_ID%20INSERT_AFTER_CURRENT_VIDEO`
2. Skip to next song: `youtubemusic://next`

**Comment by BillCipher345** [comment]
> > > > I'm sorry if I missed something, but how to do that on Windows?
> > > > 
> > > > * Interesting note - I tried to use `youtubemusic://watch?v=VIDEO_ID` - app opens, but ignores the song/video in the link.
> > > 
> > > 
> > > Because that's incorrect
> > 
> > 
> > could you please explain how is this incorrect and how to use the protocol?
> 
> Generally the protocol works like so: `youtubemusic://[method name] [...args]`. Everything after the protocol should be url encoded (i.e. `addSongToQueue ID` -> `addSongToQueue%20ID`).
> 
> Examples: `youtubemusic://play` `youtubemusic://muteUnmute` `youtubemusic://addSongToQueue%20SOME_VIDEO_ID%20INSERT_AFTER_CURRENT_VIDEO`
> 
> Here you can find all the methods available (some of them don't work for me though):
> 
> [pear-desktop/src/providers/song-controls.ts](https://github.com/pear-devs/pear-desktop/blob/b37618779eec5cfd42d2b7b12c1761c95874aa11/src/providers/song-controls.ts#L39)
> 
> Line 39 in [b376187](/pear-devs/pear-desktop/commit/b37618779eec5cfd42d2b7b12c1761c95874aa11)
> 
>  export const getSongControls = (win: BrowserWindow) => { 
> If you want to play a video/song, the (to my knowledge) best approach would be a combination of these:
> 
> 1. Add Song to Queue: `youtubemusic://addSongToQueue%20SONG_ID%20INSERT_AFTER_CURRENT_VIDEO`
> 2. Skip to next song: `youtubemusic://next`

Thank you! I will try using a redirector browser extension that will automatically do this whenever i enter a ytm link.

**Comment by BillCipher345** [comment]
Ok i managed to make it redirect ytm links to make them next in queue, but i didnt manage to find a way to automatically do `youtubemusic://next` afterwards, do u have any idea how to do this? Im using LibreWolf

**Comment by BillCipher345** [comment]
I DID IT (with the help of chat gpt)!!!! i made a custom url protocol that a redirector browser extension redirects to, which opens an ahk script that runs the `youtubemusic;//` protocol commands (i needed to tweak the timings and such a bit or else it didnt work, i made it first just open the app regularly and then open the urls).

**Comment by jwols27** [comment]
> I DID IT (with the help of chat gpt)!!!! i made a custom url protocol that a redirector browser extension redirects to, which opens an ahk script that runs the `youtubemusic;//` protocol commands (i needed to tweak the timings and such a bit or else it didnt work, i made it first just open the app regularly and then open the urls).

Could you share it, please?

**Comment by BillCipher345** [comment]
> > I DID IT (with the help of chat gpt)!!!! i made a custom url protocol that a redirector browser extension redirects to, which opens an ahk script that runs the `youtubemusic;//` protocol commands (i needed to tweak the timings and such a bit or else it didnt work, i made it first just open the app regularly and then open the urls).
> 
> Could you share it, please?

Um i think i can, but only in a few hours. And it will be a bit long but ill try.

**Comment by BillCipher345** [comment]
@jwols27 ok so first thing, you need to get this browser extension (any redirector would work but thats what i used): https://einaregilsson.com/redirector/

after u did so, u need to make a redirect rule that looks like this:

<img width="720" height="366" alt="Image" src="https://github.com/user-attachments/assets/42afbac6-176e-48e7-a9fe-6e6dcbac7158" />

now comes the more complicated part:

1. install AutoHotkey (v2) if u dont have it already
2. Create a new ahk script (name it however u want, and place it somewhere safe) and paste this code:
```
#Requires AutoHotkey v2.0

if (A_Args.Length = 0) {
    MsgBox "No URL argument received."
    ExitApp
}

fullUrl := A_Args[1]

; Remove everything after '&' if it exists
ampPos := InStr(fullUrl, "&")
if (ampPos) {
    fullUrl := SubStr(fullUrl, 1, ampPos - 1)
}

; Remove protocol
songId := StrReplace(fullUrl, "pearmusic://")

; Remove trailing slash if it exists
if (SubStr(songId, -1) = "/") {
    songId := SubStr(songId, 1, -1)
}

; Trim just in case
songId := Trim(songId)

; Debug (optional)
; MsgBox "Song ID: " songId

; Closes current tab
; SendInput "^w"

; Step 1: Open yt music
Run("C:\Users\USERNAME\AppData\Local\Programs\youtube-music\YouTube Music.exe")

Sleep(4000)

; Step 2: Add to queue
Run("youtubemusic://addSongToQueue%20" songId "%20INSERT_AFTER_CURRENT_VIDEO")

Sleep(1000)

; Step 3: Skip to it
Run("youtubemusic://next")
```

make sure to replace the `USERNAME` with your user in the path. all the ; parts are for debug and such, dont mind them, they are disabled.

3. create a new .reg file (name it however u like) and paste this code (replace paths to correct paths):
```
Windows Registry Editor Version 5.00

[HKEY_CLASSES_ROOT\pearmusic]
@="URL:pearmusic Protocol"
"URL Protocol"=""

[HKEY_CLASSES_ROOT\pearmusic\shell\open\command]
@="\"-Path to AutoHotkey64.exe" \"Path to AHK script from ealier" \"%1\""
```

when replacing paths, make sure to use double \ between folder names, and end with a single \ (idk why its like that but thats how it is).

4. its basically done now, i just put another ahk script to quickly open and close ytm app at startup in order to fix some bugs that happen when opening the app for the first time after an update. Press win+r and type `shell:startup` it should open the startup folder. now inside that folder create a new ahk script, name it however u like. and paste this code in:
```
#Requires AutoHotkey v2.0

Run("C:\Users\USERNAME\AppData\Local\Programs\youtube-music\YouTube Music.exe")

Sleep(7000)

WinClose("ahk_exe YouTube Music.exe")
```
make sure to replace the `USERNAME` with your user in the path, this script quickly runs the app and closes it on system startup.


Now enter a ytm link with your browser (https://music.youtube.com/watch?v=Iq7j441JMbA for example, peak song btw) and see if it opens the app. u might need to let it permission to always open in AutoHotkey, but after u did so it should just do it automatically.
What we did here basically is we created a new url protocol, which the redirector extension redirects to, this url protocol opens the ahk script that we created, and this script runs the youtubemusic:// url protocol, and automatically adds the song to next in queue and skips to it.
And if u want to disable it, u can just disable the redirector browser extension.

Thats it, i know its a bit long and that there are probably better approaches to this, but thats how i did it.

**Comment by jwols27** [comment]
I'm on Mac and I couldn't get it to work unfortunately. The song controls don't seem to do anything other than give focus to the app window.

I tried triggering it manually with this script but nothing happens:
```shell
#!/bin/bash

song_id="$1"

open -a "Youtube Music"
sleep 4

open "youtubemusic://addSongToQueue%20${song_id}%20INSERT_AFTER_CURRENT_VIDEO"
sleep 1

open "youtubemusic://next"
```

**Comment by BillCipher345** [comment]
> I'm on Mac and I couldn't get it to work unfortunately. The song controls don't seem to do anything other than give focus to the app window.
> 
> I tried triggering it manually with this script but nothing happens:
> 
> #!/bin/bash
> 
> song_id="$1"
> 
> open -a "Youtube Music"
> sleep 4
> 
> open "youtubemusic://addSongToQueue%20${song_id}%20INSERT_AFTER_CURRENT_VIDEO"
> sleep 1
> 
> open "youtubemusic://next"

Um i dont really know how this stuff works on mac so i cant really help u, but good luck! Maybe the url protocol works differently on mac?
